src/angular/metadataReader.ts

```diff
@@ -1,162 +1,155 @@
-      const transformed = normalizeTransformed(Config.transformTemplate(inlineTemplate.text, null, dec));
-      return {
-        template: transformed,
-        url: null,
-        node: inlineTemplate
-      };
-    } else {
-      if (external.templateUrl) {
-        try {
-          const template = this._fileResolver.resolve(external.templateUrl);
-          const transformed = normalizeTransformed(Config.transformTemplate(template, external.templateUrl, dec));
-          return {
-            template: transformed,
-            url: external.templateUrl,
-            node: null
-          };
-        } catch (e) {
-          logger.info('Cannot read the external template ' + external.templateUrl);
+
+    readComponentTemplateMetadata(dec: ts.Decorator, external: MetadataUrls): TemplateMetadata {
+        const template_M = getTemplate(dec)
+            .fmap(inlineTemplate => {
+                const transformed = normalizeTransformed(Config.transformTemplate(inlineTemplate.text, null, dec));
+                return {
+                    template: transformed,
+                    url: null,
+                    node: inlineTemplate,
+                };
+            });
+
+        if (template_M.isSomething) {
+            return template_M.unwrap();
+        } else {
+            // TODO: Refactoring this requires adding seem to fileResolver
+            if (external.templateUrl) {
+                try {
+                    const template = this._fileResolver.resolve(external.templateUrl);
+                    const transformed = normalizeTransformed(Config.transformTemplate(template, external.templateUrl, dec));
+                    return {
+                        template: transformed,
+                        url: external.templateUrl,
+                        node: null
+                    };
+                } catch (e) {
+                    logger.info('Cannot read the external template ' + external.templateUrl);
+                }
+            }
         }
-      }
     }
-  }
-
-  readComponentStylesMetadata(dec: ts.Decorator, external: MetadataUrls) {
-    const inlineStyles = getDecoratorPropertyInitializer(dec, 'styles');
-    let styles: any[];
-    if (inlineStyles && inlineStyles.kind === kinds.ArrayLiteralExpression) {
-      inlineStyles.elements.forEach((inlineStyle: any) => {
-        if (isSimpleTemplateString(inlineStyle)) {
-          styles = styles || [];
-          styles.push({
-            style: normalizeTransformed(Config.transformStyle(inlineStyle.text, null, dec)),
-            url: null,
-            node: inlineStyle,
-          });
+
+    readComponentStylesMetadata(dec: ts.Decorator, external: MetadataUrls) {
+        let styles: any[];
+        const inlineStyles_M = getInlineStyle(dec)
+            .fmap(inlineStyles => {
+                return inlineStyles.elements.map((inlineStyle: ts.Expression) => {
+                    if (isSimpleTemplateString(inlineStyle)) {
```

Comment:
Maybe we can first apply `filter` and after that map it, so we can drop the ugly if condition?